File: cpp/medium/MinimumTotal.cpp
#include <algorithm>
#include <vector>
using namespace std;

int minimumTotal(vector<vector<int>> &triangle) {
  for (int i = 1; i < triangle.size(); ++i) {
    for (int k = 0; k <= i; ++k) {
      if (k == 0) {
        triangle[i][k] += triangle[i - 1][k];
        continue;
      }
      if (k == i) {
        triangle[i][k] += triangle[i - 1][k - 1];
        continue;
      }
      triangle[i][k] += min(triangle[i - 1][k], triangle[i - 1][k - 1]);
    }
  }
  int rs = triangle.back()[0];
  for (auto i : triangle.back()) {
    rs = min(rs, i);
  }
  return rs;
}

File: go/middle/minimumTotal.go
package main

import "math"

func minimumTotal(triangle [][]int) int {
	dp := make([]int, len(triangle[len(triangle)-1]))
	for i := range triangle {
		for k := len(triangle[i]) - 1; k >= 0; k-- {
			if k == 0 {
				dp[k] = dp[k] + triangle[i][k]
			} else if k == len(triangle[i])-1 {
				dp[k] = dp[k-1] + triangle[i][k]
			} else {
				dp[k] = min(dp[k-1], dp[k]) + triangle[i][k]
			}
		}
	}
	rs := math.MaxInt
	for i := range dp {
		rs = min(rs, dp[i])
	}
	return rs
}
